### DISCUSSION TITLE
Per-window struts

### DISCUSSION THREAD

**Discussion by Nairou** [YaLTeR/niri#809] [Ideas]
I keep thinking it would be nice to be able to have different struts for my two monitors.

Use case: I have a vertical status bar on one of my windows, but not both (just the secondary window, so I can use the full space of the primary window). Niri handles the status bar fine, and reduces the available screen width. My status bar design mimics that of a regular window, so it is floating with a margin around it, so I can see windows that scroll behind it.

However, I also use left/right struts, in order to peek at neighboring windows. This works great everywhere _except_ next to the vertical bar. The gap that was a nice width for the edge of the screen looks much too wide between the current window and the status bar. It would be nice to adjust the left strut to be smaller, just for that monitor, to fit in better.

**Comment by YaLTeR** [comment]
Could you show a screenshot? I'm not sure I can visualize the mismatch you're talking about.

Also, per-output layout options is what you seem to need, and it has a feature request here: #401

**Reply by Nairou** [reply]
> Also, per-output layout options is what you seem to need, and it has a feature request here: #401

I searched but didn't see that request. That would do it, thanks!

> Could you show a screenshot? I'm not sure I can visualize the mismatch you're talking about.

Sure, here you go. Note the large gap between the left bar and the window. Without a strut, it would be the same as the top and bottom gaps.

![image](https://github.com/user-attachments/assets/65135dc3-b666-4dc2-961b-64bea528bc6a)

**Reply by YaLTeR** [reply]
Could you remove the right margin from the bar itself? So that the gap comes only from the niri strut.

**Reply by Nairou** [reply]
> Could you remove the right margin from the bar itself? So that the gap comes only from the niri strut.

The bar has no margin at all. It has a fixed position and size, but that's it. If I set left strut to zero, the gap between bar and window drops to normal (identical to other window gaps).

**Reply by YaLTeR** [reply]
The bar has a margin. Niri doesn't add any extra margins other than struts.

**Reply by Nairou** [reply]
Here it is with no struts. Only space is the layout gap (`gaps 16`).

![image_2024-11-18_10-24-58](https://github.com/user-attachments/assets/89c150cf-d9ec-48bd-aac5-21e29a620162)

And here is with `gaps 0`, again no struts.

![image_2024-11-18_10-25-27](https://github.com/user-attachments/assets/d6b6639e-7f50-4bb2-8834-2d22459d5f31)

I'm assuming niri is treating the bar as a regular window for purposes of gap separation.

**Reply by YaLTeR** [reply]
Ah, yeah. Sorry. The gaps are indeed added on the very left and right. Why do you need the extra left/right strut then?

**Reply by Nairou** [reply]
For my main monitor, the struts are nice for seeing neighboring windows. The struts are not at all necessary next to the bar, which only exists on my secondary monitor. I imagine #401 should solve the issue. :smile:

**Reply by YaLTeR** [reply]
I see, thanks for the explanation

**Comment by Paul-Eau** [comment]
I'd also like to have this setting struts per-windows using window-rule, it would be very welcome. I also use this function to set left/right struts, in order to peek at neighboring windows quickly by mouse, but the inconvenience is that I use vertical tabs in Firefox with the ability to auto hide the vertical bar. Imposing strut{left 0} would allow me to open the vertical bar just by hovering over it. 
![Screenshot from 2024-11-19 11-06-47](https://github.com/user-attachments/assets/5cdd2592-218b-403b-a5ed-8adb999a2f0e)

**Reply by YaLTeR** [reply]
Per-window struts sounds quite strange tbh. Maybe we should have some other way to get to the off-screen window without struts?

**Reply by Paul-Eau** [reply]
I don't know, that's what seemed most obvious to me, for my use case, knowing that Firefox is always the first window in my workspace, the left-strut for quick access to adjacent windows (if it's the first window in the workspace) are not useful in my case.

**Reply by Nairou** [reply]
One (optional) generalization could be that struts are only measured when there is a neighboring off-screen window on that side. If you're viewing the left-most window, then there are none to the left of that, so don't expand the left strut.

**Reply by YaLTeR** [reply]
Then window size will change depending on whether there's a window to the left of it, which is a no-go.

**Reply by Paul-Eau** [reply]
I managed to reproduce this feature for my use case with a bash script that listens to Niri events. I think the author can modify the script to listen to which screen is focused to reduce struts. I leave it below if it can help anyone facing the same situation.

```
#!/bin/bash

CONFIG_FILE="$HOME/.config/niri/config.kdl"

update_config() {
    local value="$1"
    if [[ -f "$CONFIG_FILE" ]]; then
        sed -i "/struts {/,/}/s/left [-0-9]*/left $value/" "$CONFIG_FILE"
        echo "Configuration updated: set 'left' to $value in $CONFIG_FILE"
    else
        echo "Error: Configuration file $CONFIG_FILE not found."
    fi
}

process_event() {
    while IFS= read -r line; do
        if [[ "$line" == *"Window focus changed"* ]]; then
            focused_window=$(niri msg focused-window)
            if echo "$focused_window" | grep -q 'App ID: "firefox"'; then
                update_config -8

            else
                update_config 16  
            fi
        fi
    done
}

niri msg event-stream | process_event

```
